packages/fxa-content-server/app/scripts/models/reliers/oauth.js

```diff
@@ -362,6 +363,53 @@ var OAuthRelier = Relier.extend({
     return this._wantsScopeThatHasKeys;
   },
 
+  /**
+   * Ensure the prompt=none can be used.
+   *
+   * @param {Account} account
+   * @throws {OAuthError} if prompt=none cannot be used.
+   * @returns {Promise<none>} rejects with an error if prompt=none cannot be used.
+   */
+  validatePromptNoneRequest(account) {
+    return Promise.resolve()
+      .then(() => {
+        if (!this._config.isPromptNoneEnabled) {
+          throw OAuthErrors.toError('PROMPT_NONE_NOT_ENABLED');
+        }
+
+        if (!this._config.isPromptNoneEnabledForClient) {
+          throw OAuthErrors.toError('PROMPT_NONE_NOT_ENABLED_FOR_CLIENT');
+        }
+
+        if (this.wantsKeys()) {
+          throw OAuthErrors.toError('PROMPT_NONE_WITH_KEYS');
+        }
+
+        if (account.isDefault() || !account.get('sessionToken')) {
+          throw OAuthErrors.toError('PROMPT_NONE_NOT_SIGNED_IN');
+        }
+
+        const requestedEmail = this.get('email');
+        if (requestedEmail && requestedEmail !== account.get('email')) {
+          throw OAuthErrors.toError('PROMPT_NONE_DIFFERENT_USER_SIGNED_IN');
+        }
+
+        // account has all the right bits associated with it,
+        // now let's check to see whether the account and session
+        // are verified. If session is no good, the promise will
+        // reject with an INVALID_TOKEN error.
+        return account.sessionVerificationStatus();
+      })
+      .then(({ emailVerified, sessionVerified, verified }) => {
+        // 1. unverified account, user must verify signup email
+        // 2. unverified email - when does this happen?
+        // 3. unverified session, user must perform 2fa on session
```

Comment:
I believe `verified == emailVerified || sessionVerified`, so your "when does this happen?" case for (2) is the same as for (1), i.e. they need to verify signup email.

Also, `! sessionVerified` does not necessary indicate that they need to do 2FA, it might be because this session was originally created in a context that needed keys, so the `sesson.mustVerify` flag is set and they need to do an email confirmation loop.